### DISCUSSION TITLE
How to use KafkaIdempotentRepository?

### DISCUSSION THREAD

**Discussion by dirooster** [apache/camel-karavan#1467] [Q&A] [answered]
### Ask a question

How can I use KafkaIdempotentRepository as a Idempotent Consumer?
In EIP Idempotent Consumer in Karavan we have only 5 Idempotent Repositories, and does't have Kafka. Also If I try to create custom bean with Camel, I couldn't find necessary import. 
Is it possible in Karavan? 
<img width="1792" alt="Снимок экрана 2024-12-12 в 11 11 15" src="https://github.com/user-attachments/assets/bb1afd53-8095-4bae-b028-63660a762741" />

![Uploading Снимок экрана 2024-12-12 в 17.55.43.png…]()

### Variant

None

### Container Management (if applicable)

Docker

### Operating System (if applicable)

Linux

### Version

4.8.0

**Comment by dirooster** [comment]
@davsclaus, could you help with any information?

**Comment by dirooster** [comment, marked as answer]
I got by myself 
camel.jbang.dependencies=camel-kafka

**Reply by davsclaus** [reply]
Yes you need to add the dependency because your java code is creating the bean and adding to registry manually.

**Reply by dirooster** [reply]
![Uploading Снимок экрана 2024-12-12 в 17.55.43.png…]()

@davsclaus , Hi, thank you for answer. Could you help with this problem? 
I tried different ways, but it doesn't work. How could I bind it? 

added second scrin in description

**Reply by dirooster** [reply]
import org.apache.camel.BindToRegistry;
import org.apache.camel.Configuration;
import org.apache.camel.Exchange;
import org.apache.camel.Processor;
import org.apache.camel.PropertyInject;
import org.apache.camel.processor.idempotent.kafka.KafkaIdempotentRepository;
import org.apache.camel.support.SimpleRegistry;
import org.apache.camel.impl.DefaultCamelContext;
import org.springframework.context.annotation.Bean;
import org.springframework.stereotype.Component;


@Component("KafkaIdempotentRepositoryCustom")
@BindToRegistry("KafkaIdempotentRepositoryCustom")
public class CreateCustomIdempotentRepository {

    public KafkaIdempotentRepository kir;

    public CreateCustomIdempotentRepository() {
        kir = new KafkaIdempotentRepository("idempotent-inserts", "localhost:9091");
    }

    // @Override
    // public void process(Exchange exchange) throws Exception {
    //     KafkaIdempotentRepository kafkaIdempotentRepository = new KafkaIdempotentRepository("idempotent-inserts", "localhost:9091");
    //     SimpleRegistry registry = new SimpleRegistry();
    //     registry.bind("KafkaIdempotentRepositoryCustom", kafkaIdempotentRepository);
    //     exchange.getContext().getCamelContextExtension().setRegistry(registry);
    // }

}



Caused by: org.apache.camel.NoSuchBeanException: No bean could be found in the registry for: KafkaIdempotentRepositoryCustom of type: org.apache.camel.spi.IdempotentRepository
	at org.apache.camel.reifier.AbstractReifier.mandatoryLookup(AbstractReifier.java:151)
	at org.apache.camel.reifier.IdempotentConsumerReifier.resolveIdempotentRepository(IdempotentConsumerReifier.java:63)
	at org.apache.camel.reifier.IdempotentConsumerReifier.createProcessor(IdempotentConsumerReifier.java:38)
	at org.apache.camel.reifier.ProcessorReifier.makeProcessor(ProcessorReifier.java:893)
	at org.apache.camel.reifier.ProcessorReifier.addRoutes(ProcessorReifier.java:629)
	at org.apache.camel.reifier.RouteReifier.doCreateRoute(RouteReifier.java:231)
	... 28 more

**Comment by davsclaus** [comment]
https://issues.apache.org/jira/browse/CAMEL-21542

**Reply by dirooster** [reply]
Also I found bug, if you try to create a java class for a custom repository in EIP IdempotentConsumer, you will get the following error:
KafkaIdempotentRepositoryCustom org.apache.camel.spi.IdempotentRepository
index-VT47MQdK.js:295 Error: Template not found
    at qo.generateCode (index-VT47MQdK.js:125:11985)
    at index-VT47MQdK.js:295:85682

**Comment by dirooster** [comment]
@davsclaus , Could you explain how to fix bug with adding bean in registry, I've looked at a lot of similar problems, but I haven't found solution?
